frontend/src/lib/services/sns-sale.services.ts

```
@@ -0,0 +1,425 @@
+import type { Account } from "$lib/types/account";
+import { LedgerErrorKey } from "$lib/types/ledger.errors";
+import { assertEnoughAccountFunds } from "$lib/utils/accounts.utils";
+import { logWithTimestamp } from "$lib/utils/dev.utils";
+import { toToastError } from "$lib/utils/error.utils";
+import {
+  commitmentExceedsAmountLeft,
+  validParticipation,
+} from "$lib/utils/projects.utils";
+import { getSwapCanisterAccount } from "$lib/utils/sns.utils";
+import { isNullish } from "$lib/utils/utils";
+import type { TokenAmount } from "@dfinity/nns";
+import type { Principal } from "@dfinity/principal";
+import type { Ticket } from "@dfinity/sns/dist/candid/sns_swap";
+import type { E8s } from "@dfinity/sns/dist/types/types/common";
+import { fromDefinedNullable, fromNullable } from "@dfinity/utils";
+import { get } from "svelte/store";
+import {IcrcTransferError} from "@dfinity/ledger";
+import {ledgerErrorToToastError} from "$lib/utils/sns-ledger.utils";
+import {TxDuplicateError} from "@dfinity/nns";
+
+export interface SnsTicket {
+  rootCanisterId: Principal;
+  ticket: Ticket | undefined;
+}
+
+// TODO(sale): move to ic-js
+enum GetOpenTicketErrorType {
+  TYPE_UNSPECIFIED = 0,
+  TYPE_SALE_NOT_OPEN = 1,
+  TYPE_SALE_CLOSED = 2,
+}
+export const getOpenTicket = async ({
+  withTicket,
+  rootCanisterId,
+  certified,
+}: {
+  withTicket?: boolean; // TODO: for testing purpose only
+  rootCanisterId: Principal;
+  certified: boolean;
+}): Promise<SnsTicket | undefined> => {
+  try {
+    const identity = await getCurrentIdentity();
+    const { result } = await getOpenTicketApi({
+      identity,
+      rootCanisterId,
+      withTicket,
+      certified,
+    });
+
+    const resultData = fromDefinedNullable(result);
+
+    if ("Ok" in resultData) {
+      return {
+        rootCanisterId,
+        ticket: fromNullable(resultData.Ok.ticket),
+      };
+    }
+
+    // TODO(GIX-1271): display more details in error message
+
+    if ("Err" in resultData) {
```

This is the part that we normally do in `ic-js`.

The pattern was to raise an error there instead of returning the whole response.

---

Yep, updated 👍